roaring/roaring.go

```diff
@@ -2049,97 +2011,112 @@ func (c *Container) bitmapToRun() {
 		// find next 0
 		for current == maxBitmap && i < bitmapN-1 {
 			i++
-			current = c.bitmap[i]
+			current = bitmap[i]
 		}
 
 		if current == maxBitmap {
 
 			// bitmap[1023] == maxBitmap
-			c.runs = append(c.runs, interval16{start, maxContainerVal})
+			runs = append(runs, interval16{start, maxContainerVal})
 			break
 		}
 		currentLast := uint16(trailingZeroN(^current))
 		last = 64*i + currentLast
-		c.runs = append(c.runs, interval16{start, last - 1})
+		runs = append(runs, interval16{start, last - 1})
 
 		// pad LSBs with 0s
 		current = current & (current + 1)
 	}
-
-	c.bitmap = nil
-	c.mapped = false
+	c.setRuns(runs)
 }
 
 // arrayToRun converts from array format to RLE format.
-func (c *Container) arrayToRun() {
+func (c *Container) arrayToRun(numRuns int32) {
 	statsHit("arrayToRun")
-	c.containerType = containerRun
+	array := c.array()
+	c.typ = containerRun
+	c.mapped = false
 	// return early if empty
 	if c.n == 0 {
-		c.runs = make([]interval16, 0)
-		c.array = nil
-		c.mapped = false
+		c.setRuns(nil)
 		return
 	}
+	if numRuns == 0 {
+		numRuns = arrayCountRuns(array)
+	}
 
-	numRuns := c.arrayCountRuns()
-	c.runs = make([]interval16, 0, numRuns)
-	start := c.array[0]
-	for i, v := range c.array[1:] {
-		if v-c.array[i] > 1 {
+	runs := make([]interval16, 0, numRuns)
+	start := array[0]
+	for i, v := range array[1:] {
+		if v-array[i] > 1 {
 			// if current-previous > 1, one run ends and another begins
-			c.runs = append(c.runs, interval16{start, c.array[i]})
+			runs = append(runs, interval16{start, array[i]})
 			start = v
 		}
 	}
 	// append final run
-	c.runs = append(c.runs, interval16{start, c.array[c.n-1]})
-	c.array = nil
-	c.mapped = false
+	runs = append(runs, interval16{start, array[c.n-1]})
+	c.setRuns(runs)
 }
 
 // runToArray converts from RLE format to array format.
 func (c *Container) runToArray() {
 	statsHit("runToArray")
-	c.containerType = containerArray
-	c.array = make([]uint16, 0, c.n)
+	runs := c.runs()
+	c.typ = containerArray
+	c.mapped = false
 
 	// return early if empty
 	if c.n == 0 {
-		c.runs = nil
-		c.mapped = false
+		c.setArray(nil)
 		return
 	}
 
-	for _, r := range c.runs {
+	array := make([]uint16, c.n)
+	n := int32(0)
+	for _, r := range runs {
 		for v := int(r.start); v <= int(r.last); v++ {
-			c.array = append(c.array, uint16(v))
+			array[n] = uint16(v)
+			n++
 		}
 	}
-	c.runs = nil
-	c.mapped = false
+	if roaringParanoia {
+		if n != c.n {
+			panic("run has fewer bits set than container.n")
```

Comment:
couldn't it have more as well?

Reply:
no, because that would have been caught by the previous check a few lines up (which is where it is to avoid array/slice overflows).